node/src/testing/multi_stage_test_reactor/test_chain.rs

```diff
@@ -383,6 +426,77 @@ async fn test_joiner_at_genesis() {
         .await;
 }
 
+/// Test a node joining to a single node network
+#[tokio::test]
+async fn test_archival_sync() {
+    testing::init_logging();
+
+    const INITIAL_NETWORK_SIZE: usize = 1;
+
+    let mut rng = crate::new_rng();
+
+    // Create a chain with just one node
+    let mut chain = TestChain::new(INITIAL_NETWORK_SIZE, &mut rng).await;
+
+    assert_eq!(
+        chain.network.nodes().len(),
+        INITIAL_NETWORK_SIZE,
+        "There should be just one bonded validator in the network"
+    );
+
+    // Get the first switch block hash
+    // As part of the fast sync process, we will need to retrieve the first switch block
+    let switch_block_hash = await_switch_block(1, &mut chain.network, &mut rng)
+        .await
+        .hash();
+
+    let era_to_join = 3;
+    info!("Waiting for Era {} to end", era_to_join);
+    chain
+        .network
+        .settle_on(
+            &mut rng,
+            has_passed_by_era(era_to_join),
+            Duration::from_secs(600),
+        )
+        .await;
+
+    // Have a node join the network with that hash
+    info!("Joining with trusted hash {}", switch_block_hash);
+    let joiner_node_secret_key = Arc::new(SecretKey::random(&mut rng));
+    chain
+        .add_node(
+            false,
+            joiner_node_secret_key,
+            Some(switch_block_hash),
+            true,
+            &mut rng,
+        )
+        .await;
+
+    assert_eq!(
+        chain.network.nodes().len(),
+        2,
+        "There should be two nodes in the network (one bonded validator and one read only)"
+    );
+
+    let synchronized_era = era_to_join + 1;
+    info!("Waiting for Era {} to end", synchronized_era);
+    chain
+        .network
+        .settle_on(
+            &mut rng,
+            has_passed_by_era(synchronized_era),
+            Duration::from_secs(600),
+        )
+        .await;
+
+    chain
+        .network
+        .settle_on(&mut rng, all_have_genesis_state, Duration::from_secs(600))
```

Comment:
I think this test assumes (maybe correctly) that joining node's state will be synchronized from the highest block until Genesis so this `all_have_genesis_state` is sufficient to confirm that archival functionality works. I am not sure that's enough though, what if the strategy was to start from Genesis and work the way towards the tip? Or worse, what if the joining node downloaded only the Genesis state? I think the assertion should be that the joining node has _all_ of the blocks and _all_ of the state referenced from these blocks.

Reply:
Should be fixed in ceffeeea5489aa8d2dc50de60c30adcbac260108